iss18

xavierfav opened the pull request:
In order to prevent getting duplicates of votes from the validation form:

- Added disable of the button after submitting (client side)
- Added transaction and check if annotation_id is already in the database before saving votes (server side)

alastair commented:
Please give this pull request a descriptive title, and in the comment say that it relates to issue #18 (so that github links the two).

Sometimes people put `[WIP]` at the beginning of the pull request title (Work in progress) so that others know not to review it yet.

alastair commented on datasets/views.py:
```diff
@@ -155,20 +156,25 @@ def contribute_validate_annotations_category(request, short_name, node_id):
-                    )
+                    annotation_id = form.cleaned_data['annotation_id']
+                    check = Vote.objects.filter(created_by=request.user,
+                                               annotation_id=annotation_id)
```
this is not aligned with the line above

ffont commented:
I'm waiting for @xavierfav to add some tests before I review and merge.

ffont commented:
Is this now ready for re-review?

xavierfav commented:
It is not. I struggle with "simulating" the post request. I do not get the right response.

xavierfav commented:
Ready for review

alastair commented on datasets/tests/test_views.py:
```diff
@@ -0,0 +1,58 @@
+        
+        # post requests
+        response = self.client.post(reverse('save-contribute-validate-annotations-per-category'), data=form_data)
+        response2 = self.client.post(reverse('save-contribute-validate-annotations-per-category'), data=form_data)
```
If you're not going to use `response2`, you don't need to assign this.

alastair commented on datasets/tests/test_views.py:
```diff
@@ -0,0 +1,58 @@
+                    created_at__gt=row.created_at-timedelta(seconds=10)).count() > 1:
+                duplicate_votes.append(row)
+                
+        self.assertEquals(response.status_code, 200)
```
I'm looking at the view for this test, and see that there is nothing to return a different response code when there is an error.
We should change this so that when we return an error (https://github.com/MTG/freesound-datasets/blob/367d6f5702f0ed44ebc9b999dd3b5fb6f5a69929/datasets/views.py#L184) we also use the correct HTTP code (perhaps 400), and also add a test for this case.

ffont replied:
If this is to be changed we should handle it also in the template code.
Right now the javascript only handles the `success` case (HTTP 200, https://github.com/MTG/freesound-datasets/blob/master/templates/contribute_validate_annotations_category.html#L12). 
If we add `status=400` when the forms are not valid and there are errors, then this should be handled in an `error` case in the javascript.

alastair replied:
Additionally we should decide what to do with the disabled submit button if the request fails... do we leave it disabled? Or do we re-enable it?

alastair commented on datasets/tests/test_views.py:
```diff
@@ -0,0 +1,58 @@
+                     'comment': '',
+                     'dataset': '1',
+                    }
+        for i in range(nb_annotations):      
```
In tests you should make the case that you are trying to check as simple as possible. In this case, we should only have 12 forms if this is a strict requirement which is enforced server-side in the view.
I believe this is not the case, so my preference is to have only one form in this formset.

alastair commented on datasets/tests/test_views.py:
```diff
@@ -0,0 +1,58 @@
+            form_data['form-{0}-annotation_id'.format(i)] = str(all_annotation_object_ids[i])
+            form_data['form-{0}-vote'.format(i)] = ['1', '-1', '1'][i%3]
+            
+        # login
```
This is another comment that's not needed, because it doesn't tell us anything that we didn't already know

alastair commented on datasets/tests/test_views.py:
```diff
@@ -0,0 +1,58 @@
+
+        # checking duplicates
+        duplicate_votes = []
+        for row in Vote.objects.all():
```
Again, in our test we know exactly what we are trying to submit twice. Our check in the view is based on the user and annotation_id. In this case, we can do two checks. One after the first submission (to see if we have 1 row for this user/annotationid) and another after the second submission (to check that we still have just 1)

ffont commented on datasets/tests/test_views.py:
```diff
@@ -0,0 +1,58 @@
+
+
+class ContributeTest(TestCase):
+    fixtures = ['datasets/fixtures/initial.json']
```
However I must say that I'm happy with the full path being there, in my opinion it makes it clearer where the data comes from.

xavierfav commented:
I did the following changes:
- Cleaning non useful comments.
- Add documentation to functions in generate_fake_data.py.
- Fixed the problem with the disable/unable submit button.
- Changed the strategy of test. I do three checks: (i) after submitting the 1st form, the response is the one expected, (ii) the vote has been added to the database, (iii) after a second submission, no vote is added in the database.
- Simplified the test. One fake sound and annotation created only, one annotation vote form.
- Changed print() to logging.info() (I have not done any configuration of the logging, I don't really know what is the best here and what should be exactly done)

ffont commented on datasets/tests/test_views.py:
```diff
@@ -3,56 +3,44 @@
+        
+        # check that a second vote is not added
+        self.client.post(reverse('save-contribute-validate-annotations-per-category'), data=form_data)
+        self.assertEquals(len(Vote.objects.filter(annotation_id=annotation_object_id)), 1)
```
Maybe here you should check also if the value of the vote has changed. Eg, send vote 1 with vote value "1", send vote 2 with vote value "-1", check that stored vote is still "1".

Besides that all the rest look great to me. If you add this little fix I'll merge and deploy :)

alastair replied:
Yes, agreed.

Django also has a `.count()` method which you can call on a Query object, which will give you the same number without having to call `len`

alastair commented on datasets/management/commands/generate_fake_data.py:
```diff
@@ -4,16 +4,35 @@
+    
+    Args:
+        dataset_short_name: A string corresponding to the short name of the dataset to which the sounds will be added.
+        num_sounds: An integer corresponding to the number of fake sounds to create.
```
I don't think these comments need to be so verbose - 
```
dataset_short_name: The short name of the dataset to which the sounds will be added.
num_sounds: The number of fake sounds to create.
```

but don't worry about changing them, just keep in mind for the future.

alastair commented on templates/contribute_validate_annotations_category.html:
```diff
@@ -5,7 +5,7 @@
     <script type="text/javascript">
         function submitForm(){
-            $("#submitButton").prop('disable', true);
+            document.getElementById("submitButton").disabled = true; 
```
out of interest, why did you change this from jquery (`$("....")`) to plain javascript (`document.getElementById`)? Both are correct, but we're using jquery in other places here anyway, so its use here is not a problem

xavierfav replied:
Using the prop() jquery method did not work.
I switched to the native way of doing it

alastair commented on datasets/tests/test_views.py:
```diff
@@ -11,36 +11,34 @@ class ContributeTest(TestCase):
-            form_data['form-{0}-annotation_id'.format(i)] = str(all_annotation_object_ids[i])
-            form_data['form-{0}-vote'.format(i)] = ['1', '-1', '1'][i%3]
+                    }     
+        form_data['form-0-visited_sound'] = 'Flase'
```
False

What does this field in the form mean? Is it added to the database anywhere?

alastair commented:
Please look at your configured name and email in git: https://help.github.com/articles/setting-your-username-in-git/

xavierfav commented:
Changes done
